test/kyberGovernance.js

```diff
@@ -0,0 +1,727 @@
+
+      // can not cancel when it is executed
+      await expectRevert(governance.cancel(proposalId), 'invalid state to cancel');
+      // create generic proposal, cancel it twice
+      currentTime = new BN(await Helper.getCurrentBlockTime());
+      proposalId = await createBinaryProposal(
+        executor.address,
+        votingStrategy.address,
+        targets,
+        [0],
+        signatures,
+        calldatas,
+        withDelegatecalls,
+        currentTime,
+        currentTime,
+        'link to desc'
+      );
+      // check over grace period returns true, so proposal is expired after queued
+      await executor.setExecutionData(false, false, false, false, true, delay);
+      // queue the tx
+      await governance.queue(proposalId);
+      // make it expire
+      await Helper.mineNewBlockAfter(delay);
+
+      state = await governance.getProposalState(proposalId);
+      Helper.assertEqual(ProposalState.Expired, state, 'wrong state');
+      // can not cancel when it is expired
+      await expectRevert(governance.cancel(proposalId, {from: daoOperator}), 'invalid state to cancel');
+
+      // create generic proposal, cancel it twice
+      currentTime = new BN(await Helper.getCurrentBlockTime());
+      proposalId = await createBinaryProposal(
+        executor.address,
+        votingStrategy.address,
+        targets,
+        [0],
+        signatures,
+        calldatas,
+        withDelegatecalls,
+        currentTime,
+        currentTime,
+        'link to desc'
+      );
+      // check over grace period returns true, so proposal is expired after queued
+      await executor.setExecutionData(false, false, false, false, false, delay);
+
+      await governance.cancel(proposalId);
+      state = await governance.getProposalState(proposalId);
+      Helper.assertEqual(ProposalState.Canceled, state, 'wrong state');
+      // can not cancel when it is cancelled
+      await expectRevert(governance.cancel(proposalId, {from: daoOperator}), 'invalid state to cancel');
+    });
+
+    it('reverts binary proposal executor cancel tx failed', async () => {
+      // for binary proposal, can not cancel if it is cancelled/executed/expired
+      let delay = 30;
+      // set cancel is not allowed in executor
+      await executor.setExecutionData(false, false, true, false, false, delay);
+      let currentTime = new BN(await Helper.getCurrentBlockTime());
+      let proposalId = await createBinaryProposal(
+        executor.address,
+        votingStrategy.address,
+        targets,
+        [0],
+        signatures,
+        calldatas,
+        withDelegatecalls,
+        currentTime,
+        currentTime,
+        'link to desc'
+      );
+      let state = await governance.getProposalState(proposalId);
+      Helper.assertEqual(ProposalState.Succeeded, state, 'wrong state');
+
+      await expectRevert.unspecified(governance.cancel(proposalId, {from: daoOperator}));
```

Comment:
pls specify revert message if possible

Reply:
we don't have revert since it is failed from handle cancellation in executor, similar if the strategy handle cancellation failed

Reply:
Ok